File: go/internal/e2e/llm_inference_cancel_e2e_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package e2e

import (
	"net/http"
	"sync"
	"sync/atomic"
	"testing"
	"time"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

type llmCancellingHandler struct {
	inferenceEntered atomic.Bool
	sawAbort         atomic.Bool
	abortSeen        chan struct{}
	once             sync.Once
}

func newLlmCancellingHandler() *llmCancellingHandler {
	return &llmCancellingHandler{abortSeen: make(chan struct{})}
}

func (h *llmCancellingHandler) OnLlmRequest(req *copilot.LlmInferenceRequest) error {
	served, err := llmServiceNonInference(req)
	if err != nil {
		return err
	}
	if served {
		return nil
	}
	if !llmIsInferenceURL(req.URL) {
		return llmRespondBuffered(req, 200, http.Header{"content-type": {"application/json"}}, "{}")
	}

	// Inference: never produce a response. Wait for the runtime to cancel us,
	// recording the abort.
	llmDrainRequest(req)
	h.inferenceEntered.Store(true)
	<-req.Context.Done()
	h.sawAbort.Store(true)
	h.once.Do(func() { close(h.abortSeen) })
	// Runtime already dropped the request on cancel; the sink error is a no-op.
	_ = req.ResponseBody.Error("cancelled by upstream", "cancelled")
	return nil
}

func waitFor(t *testing.T, predicate func() bool, timeout time.Duration) {
	t.Helper()
	deadline := time.Now().Add(timeout)
	for !predicate() {
		if time.Now().After(deadline) {
			t.Fatal("waitFor timed out")
		}
		time.Sleep(50 * time.Millisecond)
	}
}

func TestLlmInferenceCancel(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	handler := newLlmCancellingHandler()
	client := newLlmClient(ctx, handler)
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}

	if _, err := session.Send(t.Context(), copilot.MessageOptions{Prompt: "Say OK."}); err != nil {
		t.Fatalf("send failed: %v", err)
	}
	waitFor(t, handler.inferenceEntered.Load, 60*time.Second)
	if err := session.Abort(t.Context()); err != nil {
		t.Fatalf("abort failed: %v", err)
	}

	select {
	case <-handler.abortSeen:
	case <-time.After(30 * time.Second):
		t.Fatal("Timed out waiting for the consumer to observe runtime cancellation")
	}
	_ = session.Disconnect()

	if !handler.inferenceEntered.Load() {
		t.Fatal("Expected the inference callback to be entered")
	}
	if !handler.sawAbort.Load() {
		t.Fatal("Expected the consumer to observe the runtime-driven cancellation")
	}
}

File: java/src/test/java/com/github/copilot/LlmInferenceCancelE2ETest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static com.github.copilot.LlmInferenceTestSupport.drainRequest;
import static com.github.copilot.LlmInferenceTestSupport.headers;
import static com.github.copilot.LlmInferenceTestSupport.isInferenceUrl;
import static com.github.copilot.LlmInferenceTestSupport.newLlmClient;
import static com.github.copilot.LlmInferenceTestSupport.respondBuffered;
import static com.github.copilot.LlmInferenceTestSupport.serviceNonInference;
import static com.github.copilot.LlmInferenceTestSupport.setupCapiAuth;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.github.copilot.rpc.MessageOptions;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.SessionConfig;

/**
 * Verifies that the consumer observes a runtime-driven cancellation of an
 * in-flight inference request (the agent turn was aborted upstream).
 */
public class LlmInferenceCancelE2ETest {

    private static E2ETestContext ctx;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    private static final class CancellingHandler implements LlmInferenceProvider {

        private final AtomicBoolean inferenceEntered = new AtomicBoolean();
        private final AtomicBoolean sawAbort = new AtomicBoolean();
        private final CountDownLatch abortSeen = new CountDownLatch(1);

        @Override
        public void onLlmRequest(LlmInferenceRequest req) throws Exception {
            if (serviceNonInference(req)) {
                return;
            }
            if (!isInferenceUrl(req.getUrl())) {
                respondBuffered(req, 200, headers("content-type", "application/json"), "{}");
                return;
            }

            // Inference: never produce a response. Wait for the runtime to cancel us,
            // recording the abort.
            drainRequest(req);
            inferenceEntered.set(true);
            req.getCancellation().join();
            sawAbort.set(true);
            abortSeen.countDown();
            // Runtime already dropped the request on cancel; the sink error is a no-op.
            try {
                req.getResponseBody().error("cancelled by upstream", "cancelled");
            } catch (Exception ignored) {
                // Best effort.
            }
        }
    }

    private static void waitFor(AtomicBoolean predicate, long timeoutMillis) throws InterruptedException {
        long deadline = System.currentTimeMillis() + timeoutMillis;
        while (!predicate.get()) {
            if (System.currentTimeMillis() > deadline) {
                throw new AssertionError("waitFor timed out");
            }
            Thread.sleep(50);
        }
    }

    @Test
    void observesRuntimeDrivenCancel() throws Exception {
        setupCapiAuth(ctx);
        CancellingHandler handler = new CancellingHandler();

        try (CopilotClient client = newLlmClient(ctx, handler)) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();

            session.send(new MessageOptions().setPrompt("Say OK.")).get(60, TimeUnit.SECONDS);
            waitFor(handler.inferenceEntered, 60_000);
            session.abort().get(30, TimeUnit.SECONDS);

            assertTrue(handler.abortSeen.await(30, TimeUnit.SECONDS),
                    "Timed out waiting for the consumer to observe runtime cancellation");
            session.close();
        }

        assertTrue(handler.inferenceEntered.get(), "Expected the inference callback to be entered");
        assertTrue(handler.sawAbort.get(), "Expected the consumer to observe the runtime-driven cancellation");
    }
}
